invisible_cities/reco/psf_functions.py

```diff
@@ -0,0 +1,141 @@
+import numpy  as np
+import pandas as pd
+
+from typing import Tuple
+from typing import List
+
+from ..     core.core_functions import shift_to_bin_centers
+from ..     core.core_functions import in_range
+from .. database                import load_db
+
+def create_psf(pos    : Tuple[np.ndarray, ...],
+               charge : np.ndarray,
+               nbins  : int,
+               ranges : List[List[float]]
+               ) -> Tuple[np.ndarray, np.ndarray, List[np.ndarray]] :
+    """
+    Computes the point-spread (PSF) function of a given dataset.
+
+    Parameters
+    ----------
+    pos    : Hits relative position. Only tested for 2D.
+    charge : Hits SiPM charge normalized to the total peak charge.
+    nbins  : The number of bins in each dimension.
+    ranges : Range of the PSF in each dimension.
+
+    Returns
+    ----------
+    psf         : Point-spread function.
+    entries     : Number of entries per bin in the PSF.:
+    bin_centers : Bin centers of the PSF.
+    """
+
+    entries, edges = np.histogramdd(pos, nbins, range=ranges, normed=False)
+    sumC   , edges = np.histogramdd(pos, nbins, range=ranges, normed=False, weights=charge)
+    with np.errstate(divide='ignore', invalid='ignore'):
+        psf = np.nan_to_num(sumC / entries)
+
+    centers = [shift_to_bin_centers(edge) for edge in edges]
+
+    return psf, entries, centers
+
+
+def add_variable_weighted_mean(df        : pd.DataFrame,
+                               varMean   : str,
+                               varWeight : str
+                               ) -> pd.DataFrame :
+    """
+    Adds the average of a variable weighted by another to a
+    grouped hits DST 'df' (grouped using groupby, by event id).
+
+    Parameters
+    ----------
+    df        : dataframe (groupby by event and npeak to do it peak by peak)
+    varMean   : variable to be averaged.
+    varWeight : variable to be uses as the weight.
+    """
+    mean, weight = df.loc[:, (varMean, varWeight)].values.T
+    df.loc[:, varMean + 'peak'] = np.average(mean, weights=weight)
+
+
+def add_empty_sensors_and_normalize_q(df          : pd.DataFrame,
+                                      var         : List[str],
+                                      ranges      : List[List[float]],
+                                      database    : pd.DataFrame
+                                      ) -> pd.DataFrame :
+    """
+    Adds empty sensors to the hDST
+
+    Parameters
+    ----------
+    df  : dataframe (groupby by event and npeak to do it peak by peak)
```

Comment:
Mmmm, is that the case? Since I'm using it with apply (called from a groupby object) I was doubtful about what to put here but I guess you're right. I'll modify it.